src/hb-ot-shape-complex-arabic-pua.hh

```diff
@@ -0,0 +1,385 @@
+  {0xF14Fu, 0xF14Fu, 0xF150u, 0xF151u}, /* U+062C ARABIC LETTER JEEM */
+  {0xF152u, 0xF152u, 0xF153u, 0xF154u}, /* U+062D ARABIC LETTER HAH */
+  {0xF155u, 0xF155u, 0xF156u, 0xF157u}, /* U+062E ARABIC LETTER KHAH */
+  {0x0000u, 0x0000u, 0xF158u, 0xF158u}, /* U+062F ARABIC LETTER DAL */
+  {0x0000u, 0x0000u, 0xF159u, 0xF159u}, /* U+0630 ARABIC LETTER THAL */
+  {0x0000u, 0x0000u, 0xF15Au, 0xF15Au}, /* U+0631 ARABIC LETTER REH */
+  {0x0000u, 0x0000u, 0xF160u, 0xF160u}, /* U+0632 ARABIC LETTER ZAIN */
+  {0xF161u, 0xF161u, 0xF162u, 0xF162u}, /* U+0633 ARABIC LETTER SEEN */
+  {0xF163u, 0xF163u, 0xF164u, 0xF164u}, /* U+0634 ARABIC LETTER SHEEN */
+  {0xF165u, 0xF165u, 0xF166u, 0xF166u}, /* U+0635 ARABIC LETTER SAD */
+  {0xF167u, 0xF167u, 0xF168u, 0xF168u}, /* U+0636 ARABIC LETTER DAD */
+  {0xF169u, 0xF169u, 0xF169u, 0xF169u}, /* U+0637 ARABIC LETTER TAH */
+  {0xF16Au, 0xF16Au, 0xF16Au, 0xF16Au}, /* U+0638 ARABIC LETTER ZAH */
+  {0xF16Bu, 0xF16Cu, 0xF16Du, 0xF16Eu}, /* U+0639 ARABIC LETTER AIN */
+  {0xF16Fu, 0xF170u, 0xF171u, 0xF172u}, /* U+063A ARABIC LETTER GHAIN */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063B ARABIC LETTER KEHEH WITH TWO DOTS ABOVE */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063C ARABIC LETTER KEHEH WITH THREE DOTS BELOW */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063D ARABIC LETTER FARSI YEH WITH INVERTED V */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063E ARABIC LETTER FARSI YEH WITH TWO DOTS ABOVE */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+063F ARABIC LETTER FARSI YEH WITH THREE DOTS ABOVE */
+  {0x0000u, 0x0000u, 0x0000u, 0x0000u}, /* U+0640 ARABIC TATWEEL */
+  {0xF173u, 0xF174u, 0xF175u, 0xF175u}, /* U+0641 ARABIC LETTER FEH */
+  {0xF176u, 0xF177u, 0xF178u, 0xF178u}, /* U+0642 ARABIC LETTER QAF */
+  {0xF179u, 0xF179u, 0xF17Au, 0xF17Au}, /* U+0643 ARABIC LETTER KAF */
+  {0xF17Bu, 0xF17Bu, 0xF17Cu, 0xF17Cu}, /* U+0644 ARABIC LETTER LAM */
+  {0xF17Du, 0xF17Du, 0xF17Eu, 0xF17Eu}, /* U+0645 ARABIC LETTER MEEM */
+  {0xF17Fu, 0xF17Fu, 0xF1E1u, 0xF1E1u}, /* U+0646 ARABIC LETTER NOON */
+  {0xF1A1u, 0xF1A2u, 0xF1A3u, 0xF1A4u}, /* U+0647 ARABIC LETTER HEH */
+  {0x0000u, 0x0000u, 0xF1A5u, 0xF1A5u}, /* U+0648 ARABIC LETTER WAW */
+  {0x0000u, 0x0000u, 0xF1ABu, 0xF1ACu}, /* U+0649 ARABIC LETTER ALEF MAKSURA */
+  {0xF1A6u, 0xF1A6u, 0xF1A7u, 0xF1A8u}, /* U+064A ARABIC LETTER YEH */
+};
+
+#define SHAPING_TABLE_PUA1_FIRST	0x0622u
+#define SHAPING_TABLE_PUA1_LAST		0x064Au
+#define SHAPING_TABLE_PUA1_LEN		(SHAPING_TABLE_PUA1_LAST - SHAPING_TABLE_PUA1_FIRST + 1)
+
+static const struct ligature_set_t ligature_table_pua1[] =
+{
+  { 0xF17Bu, {
+    { 0xF144u, 0xF1C0u }, /* LAM WITH ALEF WITH MADDA ABOVE ISOLATED FORM */
+    { 0xF144u, 0xF1BEu }, /* LAM WITH ALEF WITH HAMZA ABOVE ISOLATED FORM */
+    { 0xF148u, 0xF1C2u }, /* LAM WITH ALEF WITH HAMZA BELOW ISOLATED FORM */
+    { 0xF142u, 0xF1BCu }, /* LAM WITH ALEF ISOLATED FORM */
+  }},
```

Comment:
There should be final version of these lam-alef ligatures here (the fonts has the glyphs), but the fonts use the same glyph for initial and medial lam, so I don’t think this can work in a ligature lookup. I don’t know if I can form the ligature first then add them to the fina feature (would fina be applied to a ligature made of medial + final letters)?